docs/specification.md

```diff
@@ -389,4 +398,104 @@ unit
 
 **TODO**
 
+### **lock_bond**
+
+Types
+```haskell
+type Parameter =
+  | LockBond Bond
+
+-- TODO: Not exactly sure what it is
+data Bond = Bond
+  { boMutez :: Mutez
+  , boAssets :: _undefined
+  , boPercentage :: _undefined
+  }
+
+```
+
+Parameter (in Michelson):
+```
+TODO
+```
+
+- Sender MUST submit assets/tez according to bond requirement
+
+- TODO: Explain what is the "percentage associated with bond"
+```
+if there is a percentage associated with that bond.
+Bonds are required - if there’s no value, it can be set to 0.
+```
+
+### **propose**
+
+Types
+```haskell
+type Parameter =
+  | Propose Proposal
+
+data Proposal = Proposal
+  { prName :: String
+  , prDescription :: String
+  , prMetadata :: ProposalMetaData
+  , prFrozenTokenRequirement :: Int
+  , ... -- TODO: Not sure what else
+  }
+
+data ProposalMetaData = _Hole
+```
+
+Parameter (in Michelson):
+```
+TODO
+```
+
+- Sender MUST have BOND locked as specified by requirement.
+
+### **proposal_metadata**
+
+Types
+```haskell
+type Parameter =
+  | ProposalMetadata MetadataParam
+
+data MetadataParam = MetadataParam
+  { voProposalId :: Int
```

Comment:
I think we need an identifier for a proposal so that we can vote on it. This is a placeholder for now, I'm not sure to compute this as well. What's your advice on this, Ivan?

Reply:
I think there are many options, and I am not sure which one is better. Some options I have in mind:
1. Keep it as `Int` and require to pass proposal id when `propose` is called. Then we should check whether a proposal with this ID exists. We can change `Int` to `string` or `bytes` to make proposal IDs more descriptive, the check will remain the same.
2. Keep it as `Int` and automatically assign proposal IDs. When a new proposal is made, we assign it ID of the last proposal plus 1.
3. Use `bytes` computed by hashing a proposal. This way proposal ID computation will be stateless, which is probably good (if you have a proposal, you can compute its ID without any extra data). Note that we still need to check whether such a proposal exists. Two different proposals should have different IDs, but someone may propose two equal proposals I guess (in general case). Or we can hash not only a proposal but also a counter (either a global counter or a counter per address and also hash `SENDER`). This way proposals IDs will always be different, but the computation won't be stateless anymore.

I think we can pick any of these options and keep others in mind in case the one pick appears to be not appropriate.

Reply:
I prefer the option 2 since it is the simplest. Should I mention this in the spec as well or not?